### ISSUE TITLE
ValueError: Dimension 1 in both shapes must be equal, but are 324 and 8. COCO.py

### ISSUE THREAD

**Issue by Abduoit** [matterport/Mask_RCNN#363]
Hi

Thanks for your effort, 

I tested `demo.ipynb` and it works properly. But when I change the file `mask_rcnn_coco.h5` with my own `.h5` that I trained, it gives me the following error.

Any help please ??
`
ValueError: Dimension 1 in both shapes must be equal, but are 324 and 8. Shapes are [1024,324] and [1024,8]. for 'Assign_682' (op: 'Assign') with input shapes: [1024,324], [1024,8].`

**Comment by ShiningCoding** [comment]
You must have trained an two-class classifier (including background) then. Modify the number of classes somewhere.

**Comment by Abduoit** [comment]
HI @ShiningCoding 

In coco.py there is 
```
# Number of classes (including background)
NUM_CLASSES = 1 + 80 # COCO has 80 classes
```
For me when I trained it with only one class I worte like this 

```
# Number of classes (including background)
NUM_CLASSES = 1 + 1 # COCO has 1 class
```
Is this correct ? is this what do you mean ?

**Comment by ShiningCoding** [comment]
Yes, You can try it ( in your own config class in fact). @Abduoit

**Comment by Abduoit** [comment]
@ShiningCoding 

yes that what I did, I trained it with 1 class + 1 for background 

is this what you mean ? could you please explain more ?

**Comment by tsly123** [comment]
Hi,
Although I change the `NUM_CLASSES = 1+1` with same `mask_rcnn_coco.h5`, I got the similar error as @Abduoit. Is there anything else to modified?

**Comment by Abduoit** [comment]
@tsly123 
you get this error when you use your own trained file `.h5` but what error do you get if you try to replace `mask_rcnn_coco.h5` with `mask_rcnn_balloon.h5`

please check my issue 
[here](https://github.com/matterport/Mask_RCNN/issues/366) and let me know if you are getting same error ??!!

**Comment by Venky-Iyer** [comment]
Hi.. I am facing a similar issue. When I try loading the mask_rcnn_coco.h5 weights, I get the following error:

ValueError: Dimension 1 in both shapes must be equal, but are 324 and 264. Shapes are [1024,324] and [1024,264]. for 'Assign_682' (op: 'Assign') with input shapes: [1024,324], [1024,264].

However, when I load the mask_rcnn_balloon.h5 weights, I get the following error:

ValueError: Dimension 1 in both shapes must be equal, but are 324 and 8. Shapes are [1024,324] and [1024,8]. for 'Assign_682' (op: 'Assign') with input shapes: [1024,324], [1024,8].

Can someone help.

I have 66 classes (including background) and have defined the correct number of classes in both the config classes i.e. CocoConfig() and InferenceConfig(). I have set NUM_CLASSES = 66.

**Comment by Abduoit** [comment]
@searchforpassion 
do you mean, I have to re-train it again but with removing these lines from the file `balloon.py` 
```
print("Loading weights ", weights_path)
if args.weights.lower() == "coco":
    # Exclude the last layers because they require a matching
    # number of classes
    model.load_weights(weights_path, by_name=True, exclude=[
        "mrcnn_class_logits", "mrcnn_bbox_fc",
        "mrcnn_bbox", "mrcnn_mask"])
else:
    model.load_weights(weights_path, by_name=True)
```

**Comment by xjNiu** [comment]
you can load weights with this code:
`model.load_weights(weights_path, by_name=True, exclude=[
        "mrcnn_class_logits", "mrcnn_bbox_fc",
        "mrcnn_bbox", "mrcnn_mask"])`

**Comment by xjNiu** [comment]
you can load weights with this code:

`
model.load_weights(weights_path, by_name=True, exclude=[
        "mrcnn_class_logits", "mrcnn_bbox_fc",
        "mrcnn_bbox", "mrcnn_mask"])
`

**Comment by xjNiu** [comment]
you can load weights with this code:

> model.load_weights(weights_path, by_name=True, exclude=[ "mrcnn_class_logits", "mrcnn_bbox_fc", "mrcnn_bbox", "mrcnn_mask"])

**Comment by gynr** [comment]
Changing this
```
# Number of classes (including background)
NUM_CLASSES = 1 + 80 # COCO has 80 classes
```
to this 
```
# Number of classes (including background)
NUM_CLASSES = 1 + 1 # COCO has 1 class
```
does not help. Any other solution?

**Comment by austinmw** [comment]
Did you train your own h5 file with binary class? If not exclude the fully connected layer at the end when loading the weights

**Comment by PratibhaT** [comment]
I trained for 3 classes for my own data. But while evaluating I also get a similar error. Any solution for this?

ValueError: Dimension 1 in both shapes must be equal, but are 16 and 12 for 'Assign_682' (op: 'Assign') with input shapes: [1024,16], [1024,12].

**Comment by xjNiu** [comment]
> @xjNiu 你可以具体说一下 你是如何该的那段段代码么

"Load weights"

**Comment by sheep1234567** [comment]
> you can load weights with this code:
> 
> > model.load_weights(weights_path, by_name=True, exclude=[ "mrcnn_class_logits", "mrcnn_bbox_fc", "mrcnn_bbox", "mrcnn_mask"])

@xjNiu 
Hi, thanks xjNiu! I put in this code and it works.
But the segmentation results seems wrong because I got wrong output like this:
<img width="597" alt="image" src="https://user-images.githubusercontent.com/48812213/55656937-960e7780-57ad-11e9-96c4-fca4b981fcca.png">

Is this some problems about weights or prediction code?

<img width="751" alt="image" src="https://user-images.githubusercontent.com/48812213/55656977-b0485580-57ad-11e9-8a66-86405bdac1ed.png">

**Comment by mehdiselbi** [comment]
i run the same model in two environnements (my local env and waleedka container ) and it work perefectly in the container but for my local env it raise this issue

**Comment by mehdiselbi** [comment]
replacing the model load with model.load_weights(weights_path, by_name=True, exclude=[
        "mrcnn_class_logits", "mrcnn_bbox_fc",
        "mrcnn_bbox", "mrcnn_mask"])
but it doesn't detect anything

**Comment by pkprabhat** [comment]
# create LSTM
Xi = np.reshape(X_input, (n_patterns, m_features,1))
model = Sequential()
model.add(LSTM(100, input_shape=(Xi.shape[1],Xi.shape[2]), return_sequences=True))
model.add(Dropout(0.2))
model.add(LSTM(25, activation='relu', return_sequences=True))
model.add(Dropout(0.2))
model.add(LSTM(18, activation='relu', return_sequences=False))
model.add(Dropout(0.2))
model.add(Dense(len(mini)+len(minycoord)*2, activation='sigmoid'))

# load the network weights
filename = "_five_w24X25x18X21-sigmoid-18761-0.0481-v5.0.hdf5"
model.load_weights(filename)
model.compile(loss='mean_squared_error', optimizer='adam')

This code runs well and does not give any error with 4 layer LSTM N/W. However, if I increase the neurons in middle layers as 45 and 35 respectively and use newly trained weights as below, it gives valueError.
# load the network weights
filename = "_five_w24X45x35X21-sigmoid-19459-0.0225-v5.0.hdf5"
model.load_weights(filename)
model.compile(loss='mean_squared_error', optimizer='adam')

Here's the main error message:
ValueError: Dimension 1 in both shapes must be equal, but are 100 and 180. Shapes are [100,100] and [100,180]. for 'Assign_24' (op: 'Assign') with input shapes: [100,100], [100,180].

I'm stuck with this... please explain step by step

**Comment by pedrosouzax** [comment]
Solved!

I was doing a silly thing. I was putting the path of my weights in the comand line, but the correct is put just the word "coco". This way:

python3 talhao.py train --dataset=/path/to/talhao/dataset --weights=coco

**Comment by jonasdv97** [comment]
> you can load weights with this code:
> 
> > model.load_weights(weights_path, by_name=True, exclude=[ "mrcnn_class_logits", "mrcnn_bbox_fc", "mrcnn_bbox", "mrcnn_mask"])

this will not use the weights, it will randomly initiate it, giving different predictions every time you run it.

**Comment by pra-dan** [comment]
I suffered from the same error, actually, in the `mrcnn_to_trt_single.py` file, that I used to convert my h5 model to UFF. Actually I had trained the model for 3 classes and in the `mrcnn_to_trt_single` file I had wrongly defined the `NUM_CLASSES`. And that's what lead to this. 

To trace this, I started with `NUM_CLASSES = 1 + 1` and then went on increasing the number of classes and observing the output logs. 

Classes | Output Log
--|--
|NUM_CLASSES = 1 + 1 | `ValueError: Dimension 1 in both shapes must be equal, but are 8 and 16. Shapes are [1024,8] and [1024,16]. for 'Assign_682' (op: 'Assign') with input shapes: [1024,8], [1024,16].`|
|NUM_CLASSES = 1 + 2 | `ValueError: Dimension 1 in both shapes must be equal, but are 12 and 16. Shapes are [1024,12] and [1024,16]. for 'Assign_682' (op: 'Assign') with input shapes: [1024,12], [1024,16].`|
|NUM_CLASSES = 1 + 3 |  `UFF Output written to mrcnn_nchw_custom.uff UFF Text Output written to mrcnn_nchw_custom.pbtxt`|

And voila !

**Comment by casiopa** [comment]
Hi @pra-dan.
Like you I have 3 clases. But my problem is that the other shape has this dimension [1024, 324]. I imagine that's the shape for the coco's classes (81). This is my error:

`Dimension 1 in both shapes must be equal, but are 16 and 324 for 'Assign_1365' (op: 'Assign') with input shapes: [1024,16], [1024,324].`

I'am completely stuck. Do you know what could I be doing wrong?
Thanks!!

**Comment by pra-dan** [comment]
@casiopa Are you trying to convert your model ? I mean what are you trying to do before you get stuck here ?

What I understand from your problem is that your model was trained on 81 classes (I am quite sure about that...) and NOT 4 classes. Can you ensure that first ?

**Comment by casiopa** [comment]
Thanks @pra-dan

I finally found the problem. The first time that I train with my images and classes defined, I should exclude the weights that come with the model, like this:

`model.load_weights(weights_path, by_name=True, exclude=[ "mrcnn_class_logits", "mrcnn_bbox_fc", "mrcnn_bbox", "mrcnn_mask"])`

**Comment by rooroo224** [comment]
@casiopa wow! that's incredible that you figure it out. How did you find that out? Thanks a lot!! finally, it works now
